exchange/exchange.go

```diff
@@ -305,6 +312,131 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
 	return bidResponse, err
 }
 
+func applyCategoryMapping(requestExt openrtb_ext.ExtRequest, seatBids *map[openrtb_ext.BidderName]*pbsOrtbSeatBid, categoriesFetcher stored_requests.CategoryFetcher) (map[string]string, error) {
+	res := make(map[string]string)
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if requestExt.Prebid.Targeting == nil {
+		return res, nil
+	}
+	brandCatExt := requestExt.Prebid.Targeting.IncludeBrandCategory
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if brandCatExt == (openrtb_ext.ExtIncludeBrandCategory{}) {
+		return res, nil //if not present continue the existing processing without CE.
+	}
+
+	//if includebrandcategory present but primaryadserver/publisher not present then error out the request right away.
+	primaryAdServer, err := getPrimaryAdSetver(brandCatExt.PrimaryAdServer) //1-Freewheel 2-DFP
+	if err != nil {
+		return res, err
+	}
+
+	publisher := brandCatExt.Publisher
+
+	seatBidsToRemove := make([]openrtb_ext.BidderName, 0)
+
+	for bidderName, seatBid := range *seatBids {
+		bidsToRemove := make([]int, 0)
+		for bidInd := range seatBid.bids {
+			bid := seatBid.bids[bidInd].bid
+			var tempRespBidExt ResponseBidExt
+			if len(bid.Ext) > 0 {
+				fmt.Println(string(bid.Ext))
+				if err := json.Unmarshal(bid.Ext, &tempRespBidExt); err != nil {
+					return res, err
+				}
+				//if the bid already consists bid.ext.prebid.category then it implies that adapter has returned the final Ad sever category.
+				//if not then look at the bid.cat[0] and translate it to the mapped Ad server category
+				if len(tempRespBidExt.Prebid.Targeting.HbPbCatDur) != 0 {
+					continue
+				}
+			}
+
+			bidIabCat := bid.Cat
+			if len(bidIabCat) != 1 {
+				//on receiving bids from adapters if no unique IAB category is returned  or if no ad server category is returned discard the bid
+				bidsToRemove = append(bidsToRemove, bidInd)
+			} else {
+				//if unique IAB category is present then translate it to the adserver category based on mapping file
+
+				//if mapping required but no mapping file is found then discard the bid
+				fmt.Println(primaryAdServer, publisher)
+
+				cat, err := categoriesFetcher.FetchCategories(primaryAdServer, publisher, bidIabCat[0])
+				if err != nil {
+					//if mapping required but no mapping file is found then discard the bid
+					bidsToRemove = append(bidsToRemove, bidInd)
+				}
+
+				finalCategory := fmt.Sprintf("%.2f_%s_%ds", bid.Price, cat, 30) //TODO: duration
+				res[bid.ID] = finalCategory
+			}
+
+		}
+
+		if len(bidsToRemove) > 0 {
+			if len(bidsToRemove) == len(seatBid.bids) {
+				//if all bids are invalid - remove entire seat bid
+				seatBidsToRemove = append(seatBidsToRemove, bidderName)
+			} else {
+				bids := seatBid.bids
+				for _, bidInd := range bidsToRemove {
```

Comment:
Any bids following the bid being removed will be shifted down which will throw off the indices unless you remove them in descending order by index. I believe range over slice will always iterate through the elements in ascending order, so your bidsToRemove should get built in ascending order and you can just reverse the order here when removing them: `for i=len(bidsToRemove-1); i >= 0; i-- { bidInd := bidsToRemove[i]; ...`

Reply:
^ Note that for loop I suggested assumes len() returns a signed integer, just looked it up and it in fact does return an int so that should be fine. :)

Reply:
Good catch! Fixed!

Reply:
And explicit Unit test added

Reply:
Nice, you even fixed my typo where -1 should be outside len(). Thanks! :)